pkg/scheduler/queue/tenant_querier_assignment.go

```diff
@@ -358,3 +311,152 @@ func (tqa *tenantQuerierAssignments) shuffleTenantQueriers(tenantID TenantID, sc
 	tqa.tenantQuerierIDs[tenantID] = querierIDSet
 	return true
 }
+
+// dequeueGetNode chooses the next node to dequeue from based on tenantIDOrder and tenantOrderIndex, which are
+// shared across all nodes to maintain an O(n) (where n = # tenants) time-to-dequeue for each tenant.
+// If tenant order were maintained by individual nodes, we would end up with O(mn) (where m = # query components)
+// time-to-dequeue for a given tenant.
+//
+// tenantOrderIndex is incremented, checks if the tenant at that index is a child of the current node (it may not be,
+// e.g., in the case that a tenant has queries queued for one query component but not others), and if so, returns
+// the tenant node if currentQuerier can handle queries for that tenant.
+//
+// Note that because we use the shared  tenantIDOrder and tenantOrderIndex to manage the queue, we functionally
+// ignore each Node's individual queueOrder and queuePosition.
+func (tqa *tenantQuerierAssignments) dequeueSelectNode(node *Node) (*Node, bool) {
+	// can't get a tenant if no querier set
+	if tqa.currentQuerier == nil {
+		return nil, true
+	}
+
+	checkedAllNodes := node.childrenChecked == len(node.queueMap)+1 // must check local queue as well
+
+	// advance queue position for dequeue
+	tqa.tenantOrderIndex++
+	if tqa.tenantOrderIndex >= len(tqa.tenantIDOrder) {
+		tqa.tenantOrderIndex = localQueueIndex
+	}
+
+	// no children
+	if len(node.queueMap) == 0 || tqa.tenantOrderIndex == localQueueIndex {
+		return node, checkedAllNodes
+	}
+
+	checkIndex := tqa.tenantOrderIndex
+
+	for iters := 0; iters < len(tqa.tenantIDOrder); iters++ {
+		if checkIndex >= len(tqa.tenantIDOrder) {
+			checkIndex = 0
+		}
+		tenantID := tqa.tenantIDOrder[checkIndex]
+		tenantName := string(tenantID)
+
+		if _, ok := node.queueMap[tenantName]; !ok {
+			// tenant not in _this_ node's children, move on
+			checkIndex++
+			continue
+		}
+
+		// increment nodes checked even if not in tenant-querier map
```

Comment:
```suggestion
		// at this point we know the tenant node exists in the current node's children;
		// we increment childrenChecked regardless of whether the querier is assigned to this tenant
		// so that we know when we have checked all child nodes and can stop looking
```